src/coreclr/tools/Common/JitInterface/AsyncMethodDesc.cs

```diff
@@ -0,0 +1,142 @@
+using System.Diagnostics;
+using Internal.TypeSystem;
+
+namespace Internal.JitInterface
+{
+    /// <summary>
+    /// Represents the async-callable (CORINFO_CALLCONV_ASYNCCALL) variant of a Task/ValueTask returning method.
+    /// Mirrors the structure and usage pattern of <see cref="UnboxingMethodDesc"/>.
+    /// The wrapper should be short‑lived and only used while interacting with the JIT interface.
+    /// </summary>
+    internal sealed class AsyncMethodDesc : MethodDelegator, IJitHashableOnly
+    {
+        private readonly AsyncMethodDescFactory _factory;
+        private readonly int _jitVisibleHashCode;
+
+        public MethodDesc Target => _wrappedMethod;
+
+        public AsyncMethodDesc(MethodDesc wrappedMethod, AsyncMethodDescFactory factory)
+            : base(wrappedMethod)
+        {
+            Debug.Assert(wrappedMethod is not null);
+            Debug.Assert(wrappedMethod.IsAsync && wrappedMethod.ReturnsTaskLike());
+            _factory = factory;
+            // Salt with arbitrary constant so hash space differs from underlying method.
+            _jitVisibleHashCode = HashCode.Combine(wrappedMethod.GetHashCode(), 0x51C0A54);
+        }
+
+        public override MethodDesc GetCanonMethodTarget(CanonicalFormKind kind)
+        {
+            MethodDesc realCanonTarget = _wrappedMethod.GetCanonMethodTarget(kind);
+            if (realCanonTarget != _wrappedMethod)
+                return _factory.GetAsyncMethod(realCanonTarget);
+            return this;
+        }
+
+        public override MethodDesc GetMethodDefinition()
+        {
+            MethodDesc real = _wrappedMethod.GetMethodDefinition();
+            if (real != _wrappedMethod)
+                return _factory.GetAsyncMethod(real);
+            return this;
+        }
+
+        public override MethodDesc GetTypicalMethodDefinition()
+        {
+            MethodDesc real = _wrappedMethod.GetTypicalMethodDefinition();
+            if (real != _wrappedMethod)
+                return _factory.GetAsyncMethod(real);
+            return this;
+        }
+
+        public override MethodDesc InstantiateSignature(Instantiation typeInstantiation, Instantiation methodInstantiation)
+        {
+            MethodDesc real = _wrappedMethod.InstantiateSignature(typeInstantiation, methodInstantiation);
+            if (real != _wrappedMethod)
+                return _factory.GetAsyncMethod(real);
+            return this;
+        }
+
+        public override MethodSignature Signature
+        {
+            get
+            {
+                var wrappedSignature = _wrappedMethod.Signature;
+                var ret = wrappedSignature.ReturnType;
+                var md = (MetadataType)ret;
+                Debug.Assert(md.Namespace.SequenceEqual("System.Threading.Tasks"u8));
+                ReadOnlySpan<byte> name = md.Name;
+                TypeDesc returnType = null;
+                if (name.SequenceEqual("Task"u8) || name.SequenceEqual("ValueTask"u8))
```

Comment:
Instead of checking full names, can this just do:
```
return returnType.HasInstantiation ? returnType.Instantiation[0] : this.Context.GetWellKnownType(WellKnownType.Void);
```